core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c

```diff
@@ -844,34 +864,35 @@ wasmtime_ssp_fd_close(
 #endif
     __wasi_fd_t fd)
 {
-    // Don't allow closing a pre-opened resource.
-    // TODO: Eventually, we do want to permit this, once libpreopen in
-    // userspace is capable of removing entries from its tables as well.
-    {
-        rwlock_rdlock(&prestats->lock);
-        struct fd_prestat *prestat;
-        __wasi_errno_t error = fd_prestats_get_entry(prestats, fd, &prestat);
-        rwlock_unlock(&prestats->lock);
-        if (error == 0) {
-            return __WASI_ENOTSUP;
-        }
-    }
-
     // Validate the file descriptor.
     struct fd_table *ft = curfds;
     rwlock_wrlock(&ft->lock);
+    rwlock_wrlock(&prestats->lock);
+
     struct fd_entry *fe;
     __wasi_errno_t error = fd_table_get_entry(ft, fd, 0, 0, &fe);
     if (error != 0) {
+        rwlock_unlock(&prestats->lock);
         rwlock_unlock(&ft->lock);
         return error;
     }
 
     // Remove it from the file descriptor table.
     struct fd_object *fo;
     fd_table_detach(ft, fd, &fo);
+
```

Comment:
Added logic to close a preopened resource so [close_preopen](https://github.com/WebAssembly/wasi-testsuite/blob/main/tests/rust/src/bin/close_preopen.rs) passes.

Reply:
Not sure about ⬆️. 
my concern is can users open new files in the closed preopen-directory in their following code? Since the runtime opening the resource, it might not be right to let user code to control preopen resources.

Reply:
Hmm you're right it, closing the preopened fd won't stop the user from continuing to use any fd's derived from the preopened fd. But I'm not sure if that's an issue?

I don't have all the context but here's the [wasmtime commit](https://github.com/bytecodealliance/wasmtime/commit/67e2e57b02cd244c004379690aa56d3d5e187153) where this test was added. As stated in the commit message, wasi libc doesn't expose preopened fds to user code so the user still won't be able to close them intentionally as far as I understand.

Reply:
@lum1n0us  raised a separate PR for this, https://github.com/bytecodealliance/wasm-micro-runtime/pull/2578. Please leave a comment there if you still have a concern or suggestion, thanks!